rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+
+* `inputs`: The dependencies of the flake, as a list of flake
+  references (described below).
+
+* `outputs`: A function that, given an attribute set containing the
+  outputs of each of the input flakes keyed by their identifier,
+  yields the Nix values provided by this flake. Thus, in the example
+  above, `inputs.nixpkgs` contains the result of the call to the
+  `outputs` function of the `nixpkgs` flake, and
+  `inputs.nixpkgs.packages.fuse` refers to the `packages.fuse` output
+  attribute of `nixpkgs`.
+
+  In addition to the outputs of each input, each input in `inputs`
+  also contains some metadata about the inputs. These are:
+
+  * `outPath`: The path in the Nix store of the flake's source
+    tree. This means that you could import Nixpkgs in a more
+    legacy-ish way by writing
+
+        with import inputs.nixpkgs { system = "x86_64-linux"; };
+
+    since `nixpkgs` still contains a `/default.nix`. In this case we
+    bypass its outputs entirely and only use the flake mechanism to
+    get its source tree.
+
+  * `rev`: The commit hash of the flake's Git repository.
+
+  * `revCount`: The number of ancestors of the revision `rev`. This is
+    not available for `github:...` repositories (see below), since
+    they're fetched as tarballs rather than as Git repositories.
+
+  * `lastModified`: The commit time of the revision `rev`, in the
+    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
+    this is available for both Git and GitHub repositories, so it's
+    useful for generating (hopefully) monotonically increasing version
+    strings.
+
+  The special input named `self` refers to the outputs and source tree
+  of *this* flake.
+
+  A number of outputs have a special meaning, as discussed below.
+
+## Well-known outputs
+
+A number of outputs have a specific meaning to Nix or other tools like
+Hydra. Currently, these are:
+
+* `packages`: A set of derivations used as a default by most `nix`
+  commands. For example, `nix run nixpkgs:hello` uses the
+  `packages.hello` attribute of the `nixpkgs` flake. It cannot contain
+  any non-derivation attributes. This also means it cannot be a nested
+  set! (The rationale is that supporting nested sets requires Nix to
+  evaluate each attribute in the set, just to discover which packages
+  are provided.)
+
+* `defaultPackage`: A derivation used as a default by most `nix`
+  commands if no attribute is specified. For example, `nix run
+  dwarffs` uses the `defaultPackage` attribute of the `dwarffs` flake.
+
+* `checks`: A non-nested set of derivations built by the `nix flake
+  check` command, or by Hydra if a flake does not have a `hydraJobs`
+  attribute.
+
+* `hydraJobs`: A nested set of derivations built by Hydra.
+
+* `devShell`: A derivation that defines the shell environment used by
+  `nix dev-shell` if no specific attribute is given. If it does not
+  exist, then `nix dev-shell` will use `defaultPackage`.
```

Comment:
Is `nix dev-shell` for developing the package itself, like `nix-shell '<nixpkgs>' -A hello`, or for developing _using_ the package, like `nix-shell -p pkgconfig`?

Reply:
`nix dev-shell` is for development (hence the name), while `nix run` is for using.

So `nix dev-shell` replaces `nix-shell`, while `nix run` replaces `nix-shell -p`.

Reply:
Also `!#/usr/bin/env nix dev-shell` might not work as a shebang. We'll probably still need some top-level shortcut for it.

Reply:
> Also `!#/usr/bin/env nix dev-shell` might not work as a shebang.

It does not work. Never versions of GNU `env` has support for word splitting, but tough luck if you want to be compatible.

Reply:
https://github.com/NixOS/nix/issues/3046